go/database/mpt/verification.go

```diff
@@ -570,6 +612,67 @@ func verifyHashesStoredWithParents[N any](
 	return nil
 }
 
+// verifyContractCodes runs list of validation checks on accounts and on the
+// code file stored in the given directory. These checks include:
+// 1) Fatal checks
+// - All CodeHashes within accounts are present in the code file
+// - All CodeHashes within the code file are correct matching the contract byte-codes
+// 2) Non-fatal checks
+// - There are no extra Code Hashes not referenced by any account
+func verifyContractCodes(directory string, source *verificationNodeSource, observer VerificationObserver) error {
+	observer.Progress(fmt.Sprintf("Checking contract codes ..."))
+
+	codeFile := filepath.Join(directory, "codes.dat")
+	codes, err := readCodes(codeFile)
+	if err != nil {
+		return err
+	}
+
+	checkedHashes := make(map[common.Hash]bool)
+	check := func(acc *AccountNode) error {
+		codeHash := acc.info.CodeHash
+		// skip accounts that are not contracts
+		if codeHash == emptyCodeHash {
+			return nil
+		}
+		// no need to check the hash correctness more than once
+		if _, exists := checkedHashes[codeHash]; exists {
+			return nil
+		}
+		// check that the code hash is present in the code file
+		byteCode, exists := codes[codeHash]
+		if !exists {
+			return fmt.Errorf("hash %x is missing in code file", codeHash)
+		}
+		// mark an already checked hash
+		checkedHashes[codeHash] = true
```

Comment:
I think it should be moved up even before the previous if - if there is a contract missing in the codes, and this contract is in many accounts, it will check the same account many times

Reply:
I thought about it, but it would break the check for extra hash since the count would be incorrect.

Reply:
you can utilise the fact that you store `bool` in the map.  You can store `true` when the code hash was checked and exists and `false` when it was checked but does not exist.

Reply:
Good idea

Reply:
It actually does not mess up the verification itself. Just the count. The files codes are iterated, not the account codes.